presto/dbapi.py

```diff
@@ -41,6 +42,89 @@
+
+        :param parameters: Either a dict, or a tuple, or a string with the parameters. If the query has unnamed
+        parameter stubs like "WHERE id = %s" then a list or a tuple of parameters has to be passed in the same order.
+        If the query has named parameter stubs like "WHERE id = %(user_id)s" then a dict with the same-named keys
+        should be passed as the parameters: {'user_id': user_id, ...}
+        Array parameters are also supported: "WHERE id IN %(ids)s" and {'ids': (1, 2, 3)}
+        :return: A dict if the parameters were passed as a dict; or a tuple if they were passed as a tuple or a list
+        :raises TypeError: if anything but a dict/list/tuple is passed into parameters
+        """
+        if isinstance(parameters, dict):
+            return {k: self._escape_item(v) for k, v in parameters.items()}
+        elif isinstance(parameters, (list, tuple)):
+            return tuple(self._escape_item(x) for x in parameters)
+        else:
+            raise TypeError("Unsupported parameters type. Parameters: " + str(parameters) + " Type: " +
+                            type(parameters) + ".")
+
+    @staticmethod
+    def _escape_string(item):
+        """
+        Escapes string query parameters.
+        1. UTF-decode string if needed (to support old sqlalchemy)
+        2. Escape single quotes by additional single quotes to prevent SQL injection
+        3. Add surrounding single quotes to be inserted into SQL query later
+        The single quote is the only special character to escape in Presto, and it is escaped by another single quote
+        character. Escaping single quote in string parameters protects from semicolon ";" SQL injections, as well as
+        comment "--" injections, as well as injections with dummy filters like "' or 1=1 --".
+
+        :param str item: The string to escape
+        :return: UTF-decoded, single-quote-escaped string, surrounded by additional single quotes
+        """
+        # Need to decode UTF-8 because of old sqlalchemy.
+        # Newer SQLAlchemy checks dialect.supports_unicode_binds before encoding Unicode strings as byte strings.
+        # The old version always encodes Unicode as byte strings, which would otherwise break string formatting here.
+        if isinstance(item, bytes):
+            item = item.decode('utf-8')
+
+        return "'" + item.replace("'", "''") + "'"
+
+    def _escape_sequence(self, sequence):
+        """
+        Escapes iterable collections. Collections containing other collections will be escaped recursively through
+        _escape_item().
+
+        :param Iterable sequence: The sequence to be escaped
+        :return: String containing the escaped given collection, with parentheses to make a proper SQL array
+        """
+        items = map(str, map(self._escape_item, sequence))
+        return '(' + ", ".join(items) + ')'
+
+    def _escape_item(self, item):
+        """
+        Escapes a single item of the parameters.
```

Comment:
This seems incorrect, as NULL in SQL should be without quotes. This is transforming it into a string “NULL” (which is a valid non-null string).

Reply:
I see what you mean, and I see why this is misleading. It must be correct, however. These parameters are embedded into the SQL query in method `execute` using % operator - see line 320. And that SQL query (called `operation` there) actually IS a string in Python. So embedding a string 'NULL' into a string query is correct.

Hm, on the other hand, there can be usage inconsistency. If you write queries with `WHERE id = %(value)s` or `WHERE id IN %(values)s`, NULL doesn't work as a parameter, because NULL isn't equal to anything in SQL. The check for NULL is `IS NULL`. If we add a NULL parameter after the `IS` keyword, other parameter values won't work there. Which brings me to conclusion that checking NULL value is usually a specific enough situation to be written explicitly, without parameters. Also, SQL can be written like `WHERE phone IN %(values)s OR phone IS NULL`.
So I guess we should ignore None whatsoever, and make users move this logic to their queries. What do you think?

Reply:
Ah, I see what you mean. Thanks for explaining. How do the drivers for MySQL and PostgreSQL handle this?

Reply:
I don't have a working MySQL instance to check it in.
But I've just checked it in Postgres Python client psycopg2. Apparently it converts a Python None to NULL in the SQL query. It works in queries like `WHERE id IS %(none_param)s`, and doesn't work in queries like `WHERE id = %(none_param)s` - because NULL equality is never true in SQL standard.
Besides, I didn't find special `None` treatment in psycopg2's source code in C except converting Python's `None` to `NULL`.

See this source code file:
https://github.com/psycopg/psycopg2/blob/master/psycopg/cursor_type.c
Line 376 starts the cursor's execute method `_psyco_curs_execute`.
In the line 399, it calls function `_mogrify` which validates and prepares the query and vars (its parameters). The function `_mogrify` is lines 130-305. Lines 205-213 are of the most interest, where it says that "None (Py_None) is always converted to NULL".
After `_mogrify` call, the function `_psyco_curs_merge_query_args` is called in line 404, which is defined in lines 316-371. I didn't find any special Py_None or NULL processing in it.

All and all, my observation matches to my findings in psycopg2 source code. I guess we can do the same and keep converting None value to NULL in SQL.